docs/design/proposals/healer-controller.md

```diff
@@ -0,0 +1,43 @@
+* Achieve a way to monitor the health of volumes.
+* Ceph-CSI has healer component which take cares of restarting the userspace processes when the node plugin is restarted, but in many places it uses Kubernetes specific code, with this controller model we are moving out the Kubernetes specific code out, and help make Ceph-CSI orchestrator independent.
+* We currently do not have a way to act on failed volumes. Once the controller detects that the volume is not healthy, it will have an opportunity to act on it. For example userspace mounters such as NBD has a process running per volume, this controller should ideally help restart such processes in case if they are not-started or crashed.
+* This will be a generic controller design for any mounters added in the future, for userspace mounters it will offer healing in addition, like restarting the respective processes if the volume goes abnormal. Some example userspace mounters would be rbd-nbd, cephfs-fuse, etc.
+
+# Details
+
+                    .-------------------.
+                    |     CSI-Addons    |
+                    | Healer Controller |
+                    '-------------------'
+                              |
+                              | gRPC
+                              |
+            .---------+------------------------------.
+            |         |                              |
+            |  .------------.        .------------.  |
+            |  | CSI-Addons |  gRPC  |    CSI     |  |
+            |  |  side-car  |--------| NodePlugin |  |
+            |  '------------'        '------------'  |
+            | CSI-driver Pod                         |
+            '----------------------------------------'
+
+## CSI-Addons Healer Controller
+* The healer controller will be implemented as part of the CSI-Addons controller. It will trigger controller RPC to check the health condition of the CSI volumes.
+* Watch for the `CSIAddonsNode` object, if the object is created/modified by the side-car do below operations
+    * Once the connection is established, fetch the capabilities list and check for nodes that supports `VOLUME_HEALER` capability
+    * Fetch all the VolumeAttachment` list
+    * Filter the `VolumeAttachment` list through matching driver name, node name and status attached
+    * For each `VolumeAttachment` get the respective PersistentVolume information and check the criteria of PersistentVolume Bound (also we can have configuration options like mounter type)
+    * Send the RPC request for side-car to check if the volume is healthy along with the PersistentVolume details and may attempt to heal the volume if it is found abnormal
+    * Log the final response returned from the side-car
+
+## CSI-Addons side-car
+* When a nodeplugin pod is restarted the side-car will create [CSIAddonsNode](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/api/v1alpha1/csiaddonsnode_types.go) object.
```

Comment:
will create or update the object as per https://github.com/csi-addons/kubernetes-csi-addons/blob/57ed4f6211c0847eecead1d6f7baaee15f8e356d/sidecar/internal/csiaddonsnode/csiaddonsnode.go#L135-L138?

Reply:
I'm convinced with a comment by Niels earlier:

```
I think you are concerned that the CSI-driver container in the pod restarted, and not the pod was re-created.
Only in that case, the CSIAddonsNode object would exist. It is not clear to me if this is something that can happen in reality.
```

Reply:
if the objects already exist there should be an update operation to let the controller know that sidecar is restarted. am missing this point.

Reply:
@Madhu-1 
As far as I understand from Niels previous comments, the object will be existing only when the container is restarted and it will be absent when the whole pod is restarted. If that is the case, when do we expect the container to restart, practically? Hence is it worth mentioning it here?

Reply:
Yes, correct i don't have a practical answer for this one either. i was thinking object update might help which is not present currently.

Reply:
let's take it later? maybe based on future discussions?